x-pack/plugins/security_solution/public/common/lib/cell_actions/field_value.tsx

```diff
@@ -0,0 +1,162 @@
+import { parseValue } from '../../../timelines/components/timeline/body/renderers/parse_value';
+import { EmptyComponent, getLinkColumnDefinition } from './helpers';
+import { getField, getFieldKey } from '../../../helpers';
+import { useIsExperimentalFeatureEnabled } from '../../hooks/use_experimental_features';
+
+const useFormattedFieldProps = ({
+  rowIndex,
+  pageSize,
+  ecsData,
+  columnId,
+  header,
+  data,
+}: {
+  rowIndex: number;
+  data: TimelineNonEcsData[][];
+  ecsData: Ecs[];
+  header?: ColumnHeaderOptions;
+  columnId: string;
+  pageSize: number;
+}) => {
+  const pageRowIndex = getPageRowIndex(rowIndex, pageSize);
+  const usersEnabled = useIsExperimentalFeatureEnabled('usersEnabled');
+  const ecs = ecsData[pageRowIndex];
+  const link = getLinkColumnDefinition(columnId, header?.type, header?.linkField, usersEnabled);
+  const linkField = header?.linkField ? header?.linkField : link?.linkField;
+  const linkValues = header && getOr([], linkField ?? '', ecs);
+  const eventId = (header && get('_id' ?? '', ecs)) || '';
+  const rowData = useMemo(() => {
+    return {
+      data: data[pageRowIndex],
+      fieldName: columnId,
+    };
+  }, [pageRowIndex, columnId, data]);
+
+  const values = useGetMappedNonEcsValue(rowData);
+  const value = parseValue(head(values));
+  const title = values && values.length > 1 ? `${link?.label}: ${value}` : link?.label;
+  // if linkField is defined but link values is empty, it's possible we are trying to look for a column definition for an old event set
+  if (linkField !== undefined && linkValues.length === 0 && values !== undefined) {
+    const normalizedLinkValue = getField(ecs, linkField);
+    const normalizedLinkField = getFieldKey(ecs, linkField);
+    const normalizedColumnId = getFieldKey(ecs, columnId);
+    const normalizedLink = getLinkColumnDefinition(
+      normalizedColumnId,
+      header?.type,
+      normalizedLinkField,
+      usersEnabled
+    );
+    return {
+      pageRowIndex,
+      link: normalizedLink,
+      eventId,
+      fieldFormat: header?.format || '',
+      fieldName: normalizedColumnId,
+      fieldType: header?.type || '',
+      value: parseValue(head(normalizedColumnId)),
+      values,
+      title,
+      linkValue: head<string>(normalizedLinkValue),
+    };
+  } else {
+    return {
+      pageRowIndex,
+      link,
+      eventId,
+      fieldFormat: header?.format || '',
+      fieldName: columnId,
+      fieldType: header?.type || '',
+      value,
+      values,
+      title,
+      linkValue: head<string>(linkValues),
+    };
+  }
+};
+
+export const FieldValueCell = ({
+  data,
+  ecsData,
+  header,
+  timelineId,
+  pageSize,
+  closeCellPopover,
+}: {
+  data: TimelineNonEcsData[][];
+  ecsData: Ecs[];
+  header?: ColumnHeaderOptions;
+  timelineId: string;
+  pageSize: number;
+  closeCellPopover?: () => void;
+}) => {
+  if (header !== undefined) {
+    return function FieldValue({
```

Comment:
Same here, would be nice to pull this out of the component

Reply:
I need to leave these in here because they form a closure around `data` which is used with the props in the returned component